[Content Filtering] Avoid creating multiple WCRBrowserEngineClient instances

szewai opened the pull request:
#### 4d8b71918acdc7d53f68ea74f5c5a53fbe44d7eb
<pre>
[Content Filtering] Avoid creating multiple WCRBrowserEngineClient instances
<a href="https://bugs.webkit.org/show_bug.cgi?id=289915">https://bugs.webkit.org/show_bug.cgi?id=289915</a>
<a href="https://rdar.apple.com/147251812">rdar://147251812</a>

Reviewed by Chris Dumez and Ben Nham.

The initialization of WCRBrowserEngineClient can be slow, so we should avoid creating multiple WCRBrowserEngineClient
objects. This patch ensures a process only creates one WCRBrowserEngineClient instance by introducing
ParentalControlsContentFilter class, which is a wrapper of WCRBrowserEngineClient. ContentFilterUnblockHandler and
ContentFilter objects now always use the ParentalControlsContentFilter singleton to perform operations, instead
of their own instances.

To make the implementation more robust, instead of setting m_usesWebContentRestrictions to request using
WCRBrowserEngineClient after the creation of PlatformContentFilter object, we now set the flag at creation time, and the
flag cannot be changed after creation. This ensures no switch between implementations (WebFilterEvaluator v.s.
WCRBrowserEngineClient) after content filtering starts.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/ContentFilter.cpp:
(WebCore::ContentFilter::create):
(WebCore::ContentFilter::setUsesWebContentRestrictions): Deleted.
* Source/WebCore/loader/ContentFilter.h:
* Source/WebCore/loader/ContentFilterClient.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::startLoadingMainResource):
(WebCore::DocumentLoader::usesWebContentRestrictions):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/platform/ContentFilterUnblockHandler.h:
* Source/WebCore/platform/PlatformContentFilter.h:
(WebCore::PlatformContentFilter::setUsesWebContentRestrictions): Deleted.
* Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
(WebCore::ContentFilterUnblockHandler::requestUnblockAsync const):
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
(WebCore::NetworkExtensionContentFilter::create):
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
(WebCore::ParentalControlsContentFilter::enabled const):
(WebCore::ParentalControlsContentFilter::create):
(WebCore::ParentalControlsContentFilter::ParentalControlsContentFilter):
(WebCore::ParentalControlsContentFilter::responseReceived):
(WebCore::ParentalControlsContentFilter::addData):
(WebCore::ParentalControlsContentFilter::finishedAddingData):
(WebCore::ParentalControlsContentFilter::unblockHandler const):
(WebCore::ParentalControlsContentFilter::updateFilterState):
(WebCore::ParentalControlsContentFilter::enabled): Deleted.
(WebCore::ParentalControlsContentFilter::setUsesWebContentRestrictions): Deleted.
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h: Copied from Source/WebCore/platform/cocoa/PhotosFormatSoftLink.mm.
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm: Added.
(WebCore::wcrBrowserEngineClientEnabled):
(WebCore::ParentalControlsURLFilter::singleton):
(WebCore::ParentalControlsURLFilter::ParentalControlsURLFilter):
(WebCore::ParentalControlsURLFilter::isEnabled const):
(WebCore::ParentalControlsURLFilter::isURLAllowed):
(WebCore::ParentalControlsURLFilter::allowURL):
* Source/WebCore/platform/cocoa/PhotosFormatSoftLink.mm:
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm:
(WebCore::CDMPrivateMediaSourceAVFObjC::cdm const):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.mm:
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.mm:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm:
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.mm:
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h:
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.h:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
* Source/WebCore/platform/graphics/mac/controls/SearchControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/SearchControlMac.mm:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
* Source/WebCore/testing/MockContentFilter.cpp:
(WebCore::MockContentFilter::create):
* Source/WebCore/testing/MockContentFilter.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startContentFiltering):
(WebKit::NetworkResourceLoader::usesWebContentRestrictions):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:

Canonical link: <a href="https://commits.webkit.org/292963@main">https://commits.webkit.org/292963@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/405f04ad5da96b94e58a1716024a75b3b3300368

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97566 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17191 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7407 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102670 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48095 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99611 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17484 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25644 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100569 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13244 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/88229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54687 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13014 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6112 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47537 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6191 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104673 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24646 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83388 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84355 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82810 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20842 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27358 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/5047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18239 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24607 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24429 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27743 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26003 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
```diff
@@ -57,25 +55,20 @@ class ParentalControlsContentFilter final : public PlatformContentFilter {
-#else
-    static bool enabled();
-#endif
+    ParentalControlsContentFilter(const PlatformContentFilter::FilterParameters&);
```
`explicit`

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,99 @@
+{
+    ASSERT(isMainThread());
+
+    static NeverDestroyed<RefPtr<ParentalControlsURLFilter>> filter;
```
Should use `MainThreadNeverDestroyed` and drop the assertion above.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
```diff
@@ -0,0 +1,53 @@
+class ParentalControlsURLFilter : public RefCountedAndCanMakeWeakPtr<ParentalControlsURLFilter> {
+    WTF_MAKE_TZONE_ALLOCATED(ParentalControlsURLFilter);
+public:
+    static Ref<ParentalControlsURLFilter> defaultParentalControlsURLFilter();
```
For static analysis, it is important to follow the coding style and use the `Singleton()` naming. This should likely be named `parentalControlsURLFilterSingleton()` and should return a `ParentalControlsURLFilter&`

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
```diff
@@ -0,0 +1,53 @@
+
+namespace WebCore {
+
+class ParentalControlsURLFilter : public RefCountedAndCanMakeWeakPtr<ParentalControlsURLFilter> {
```
Why subclass `RefCounted` if this is a singleton class? We usually just have empty `ref() / deref()` function on the class.

cdumez replied:
Also, do we really need `CanMakeWeakPtr`? It's kind of weird for a singleton since it never gets destroyed. More likely we can hold a RefPtr to this object instead of a WeakPtr if the idea is to silence static analysis.

szewai replied:
I changed it to `CanMakeCheckedPtr`.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,99 @@
+
+void ParentalControlsURLFilter::isURLAllowed(const URL& url, CompletionHandler<void(bool, Vector<uint8_t>&&)>&& completionHandler)
+{
+    if (!m_wcrBrowserEngineClient)
```
Can we assert that we're on the main thread?

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,99 @@
+
+void ParentalControlsURLFilter::allowURL(const URL& url, CompletionHandler<void(bool)>&& completionHandler)
+{
+    if (!m_wcrBrowserEngineClient)
```
Can we assert that we're on the main thread?

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,99 @@
+        return completionHandler(true, { });
+
+    [m_wcrBrowserEngineClient evaluateURL:url withCompletion:makeBlockPtr([completionHandler = WTFMove(completionHandler)](BOOL shouldBlock, NSData *replacementData) mutable {
+        completionHandler(!shouldBlock, makeVector(replacementData));
```
Are we sure this always runs on the main thread? We should at least assert it.

cdumez replied:
Also, it seems it would be more efficient to call `span(replacementData)` and pass a span to the completion handler instead of a vector. We never end up needing a vector, just a span so there is no reason to copy the bytes to a vector here I think.

szewai replied:
Changed `CompletionHandler` to take `NSData*` to avoid copy.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,99 @@
+        return completionHandler(true);
+
+    [m_wcrBrowserEngineClient allowURL:url withCompletion:makeBlockPtr([completionHandler = WTFMove(completionHandler)](BOOL didAllow, NSError *) mutable {
+        completionHandler(didAllow);
```
Are we sure this always runs on the main thread? We should at least assert it.

cdumez commented on Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.h:
```diff
@@ -52,7 +52,7 @@ class CDMPrivateMediaSourceAVFObjC final : public CDMPrivateInterface, public Ca
     RefPtr<LegacyCDMSession> createSession(LegacyCDMSessionClient&) override;
 
-    LegacyCDM& cdm() const { return m_cdm.get(); }
+    LegacyCDM& cdm() const;
```
Hopefully this doesn't introduce a whole lot of safer cpp failures...

szewai replied:
The safer cpp bot seems ok in last run (this change is needed to avoid including header, or let me know if you prefer including header).

cdumez commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
```diff
@@ -27,14 +27,14 @@
 
 #if PLATFORM(MAC)
 
+#import "ControlFactoryMac.h"
```
Is this avoidable? Not great from a build time perspective.

szewai replied:
No, will get build failure.

cdumez replied:
What I meant is: Can we move the destructor out of line and keep a forward declaration?

szewai replied:
So there is no destructor in this file; the failure is from line `const CheckedRef<ControlFactoryMac> m_controlFactory;` (`error: member access into incomplete type 'WebCore::ControlFactoryMac' PtrTraits::unwrap(ptr)->decrementCheckedPtrCount();`)

cdumez replied:
There is no *explicit* destructor. Therefore there is an auto-generated *inline* destructor. I think we should define it explicitly and move it out of line. Hopefully it would allow us to forward declare.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,99 @@
+{
+    ASSERT(isMainThread());
+
+    auto& filter = globalDefaultParentalControlsURLFilter();
```
Not sure we need a separate `globalDefaultParentalControlsURLFilter()` function since it is only used here.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
```diff
@@ -0,0 +1,53 @@
+public:
+    static Ref<ParentalControlsURLFilter> defaultParentalControlsURLFilter();
+
+    bool enabled();
```
`isEnabled()`. Also, can likely be `const`

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
```diff
@@ -80,26 +90,19 @@ static inline bool canHandleResponse(const ResourceResponse& response)
             if (CheckedPtr checkedThis = weakThis.get())
-                checkedThis->updateFilterState(shouldBlock, replacementData);
-        }];
+                checkedThis->updateFilterState(!allowed, WTFMove(replacementData));
```
Ideally, `updateFilterState()` would take in a enum class. The switching back and forth between `allowed` (which should be named `isAllowed` BTW) and `shouldBlock` is error-prone.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
```diff
@@ -184,15 +187,10 @@ static inline bool canHandleResponse(const ResourceResponse& response)
 #if HAVE(WEBCONTENTRESTRICTIONS)
 
-void ParentalControlsContentFilter::updateFilterState(bool shouldBlock, NSData *replacementData)
+void ParentalControlsContentFilter::updateFilterState(bool shouldBlock, Vector<uint8_t>&& replacementData)
```
Looks like we just need a span here, not a Vector.

cdumez replied:
I'm actually not entirely clear why you went from a NSData to a Vector. It seems we get a NSData from the underlying framework and store it as an NSData in the end. It seems a bit wasteful. Switching from Vector to span would make 1 copy of the bytes instead of 2. But it looks like if we kept using `NSData` all the way through, there would be 0 copies?

szewai replied:
Changed `CompletionHandler` to take `NSData*` instead.

cdumez commented on Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
```diff
@@ -178,14 +179,13 @@
-                RELEASE_LOG(ContentFiltering, "WebFilterEvaluator %s the unblock request.\n", didAllow ? "allowed" : "did not allow");
+        Ref urlFilter = WebCore::ParentalControlsURLFilter::defaultParentalControlsURLFilter();
+        urlFilter->allowURL(*m_evaluatedURL, [decisionHandler = WTFMove(decisionHandler)](bool didAllow) {
+            callOnMainThread([decisionHandler = WTFMove(decisionHandler), didAllow]() {
```
Is it always called on a background thread? If not, we could consider calling `ensureOnMainThread()`

szewai replied:
I think the problem is the completion handler can be invoked synchronously and we want an asynchronous result (not sure if async result is actually needed but given the function name is `requestUnblockAsync` I kept this pattern as the old implementation cc @aestes).

cdumez commented on Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
```diff
@@ -178,14 +179,13 @@
-            callOnMainThread([decisionHandler, didAllow] {
-                RELEASE_LOG(ContentFiltering, "WebFilterEvaluator %s the unblock request.\n", didAllow ? "allowed" : "did not allow");
+        CheckedRef urlFilter = WebCore::ParentalControlsURLFilter::singleton();
+        urlFilter->allowURL(*m_evaluatedURL, [decisionHandler = WTFMove(decisionHandler)](bool didAllow) {
```
Why not just use `WebCore::ParentalControlsURLFilter::singleton()` here? Why the need for a CheckedRef variable for a singleton?

cdumez commented on Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
```diff
@@ -178,14 +179,13 @@
+        CheckedRef urlFilter = WebCore::ParentalControlsURLFilter::singleton();
+        urlFilter->allowURL(*m_evaluatedURL, [decisionHandler = WTFMove(decisionHandler)](bool didAllow) {
+            callOnMainThread([decisionHandler = WTFMove(decisionHandler), didAllow]() {
+                RELEASE_LOG(ContentFiltering, "ParentalControlsURLFilter %s the unblock request.\n", didAllow ? "allowed" : "did not allow");
```
Need to be using `PUBLIC_LOG_STRING`

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
```diff
@@ -48,25 +52,30 @@
-#if HAVE(WEBCONTENTRESTRICTIONS)
-bool ParentalControlsContentFilter::enabled(bool usesWebContentRestrictions)
-#else
 bool ParentalControlsContentFilter::enabled()
```
Can probably be `const`?

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
```diff
@@ -48,25 +52,30 @@
+ParentalControlsContentFilter::ParentalControlsContentFilter(const PlatformContentFilter::FilterParameters& params)
+{
+#if HAVE(WEBCONTENTRESTRICTIONS)
+    m_usesWebContentRestrictions = params.usesWebContentRestrictions;
```
This should use the initialization list.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
```diff
@@ -80,26 +89,20 @@ static inline bool canHandleResponse(const ResourceResponse& response)
         m_evaluatedURL = response.url();
         m_state = State::Filtering;
-        [m_wcrBrowserEngineClient evaluateURL:*m_evaluatedURL withCompletion:[weakThis = WeakPtr { *this }](BOOL shouldBlock, NSData *replacementData) {
+        CheckedRef urlFilter = ParentalControlsURLFilter::singleton();
```
Ditto about using a CheckedRef for a singleton..

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
```diff
@@ -184,17 +187,12 @@ static inline bool canHandleResponse(const ResourceResponse& response)
 #if HAVE(WEBCONTENTRESTRICTIONS)
 
-void ParentalControlsContentFilter::updateFilterState(bool shouldBlock, NSData *replacementData)
+void ParentalControlsContentFilter::updateFilterState(State state, NSData* replacementData)
```
Please leave the star where it was. It is correct for ObjC types.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
```diff
@@ -0,0 +1,54 @@
+
+namespace WebCore {
+
+class ParentalControlsURLFilter : public CanMakeCheckedPtr<ParentalControlsURLFilter> {
```
I don't think a singleton class needs to subclass `CanMakeCheckedPtr`. Static analysis understands lifetime when you call `singleton()`.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
```diff
@@ -57,25 +55,20 @@ class ParentalControlsContentFilter final : public PlatformContentFilter {
 #if HAVE(WEBCONTENTRESTRICTIONS)
-    void updateFilterState(bool shouldBlock, NSData *replacmentData);
-    void setUsesWebContentRestrictions(bool) final;
+    void updateFilterState(PlatformContentFilter::State, NSData*);
```
Please use `NSData *replacementData`, with the star on the right side and the typo fixed in the argument name (which I think we shouldn't omit).

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
```diff
@@ -0,0 +1,51 @@
+    static ParentalControlsURLFilter& singleton();
+
+    bool isEnabled() const;
+    void isURLAllowed(const URL&, CompletionHandler<void(bool, NSData*)>&&);
```
Star is on the wrong side for `NSData *`

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,98 @@
+
+ParentalControlsURLFilter& ParentalControlsURLFilter::singleton()
+{
+    ASSERT(isMainThread());
```
Unnecessary given that you're using `MainThreadNeverDestroyed` below.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,98 @@
+{
+    ASSERT(isMainThread());
+
+    static MainThreadNeverDestroyed<std::unique_ptr<ParentalControlsURLFilter>> filter;
```
Ideally we'd just use this:
```
static MainThreadNeverDestroyed<UniqueRef<ParentalControlsURLFilter>> filter = makeUniqueRef< ParentalControlsURLFilter>();
return filter.get();
```

I think you'll need a friend declaration in the header, like we usually do.

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,98 @@
+    ASSERT(isMainThread());
+
+    if (!m_wcrBrowserEngineClient)
+        return completionHandler(true, { });
```
Can we use `nullptr` for the second parameter, for clarity?

cdumez commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -0,0 +1,98 @@
+    return !!m_wcrBrowserEngineClient;
+}
+
+void ParentalControlsURLFilter::isURLAllowed(const URL& url, CompletionHandler<void(bool, NSData*)>&& completionHandler)
```
NSData star is on the wrong side.

cdumez reviewed (approved):
LGTM with nits.

szewai committed:
[Content Filtering] Avoid creating multiple WCRBrowserEngineClient instances
https://bugs.webkit.org/show_bug.cgi?id=289915
rdar://147251812

Reviewed by Chris Dumez and Ben Nham.

The initialization of WCRBrowserEngineClient can be slow, so we should avoid creating multiple WCRBrowserEngineClient
objects. This patch ensures a process only creates one WCRBrowserEngineClient instance by introducing
ParentalControlsContentFilter class, which is a wrapper of WCRBrowserEngineClient. ContentFilterUnblockHandler and
ContentFilter objects now always use the ParentalControlsContentFilter singleton to perform operations, instead
of their own instances.

To make the implementation more robust, instead of setting m_usesWebContentRestrictions to request using
WCRBrowserEngineClient after the creation of PlatformContentFilter object, we now set the flag at creation time, and the
flag cannot be changed after creation. This ensures no switch between implementations (WebFilterEvaluator v.s.
WCRBrowserEngineClient) after content filtering starts.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/ContentFilter.cpp:
(WebCore::ContentFilter::create):
(WebCore::ContentFilter::setUsesWebContentRestrictions): Deleted.
* Source/WebCore/loader/ContentFilter.h:
* Source/WebCore/loader/ContentFilterClient.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::startLoadingMainResource):
(WebCore::DocumentLoader::usesWebContentRestrictions):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/platform/ContentFilterUnblockHandler.h:
* Source/WebCore/platform/PlatformContentFilter.h:
(WebCore::PlatformContentFilter::setUsesWebContentRestrictions): Deleted.
* Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
(WebCore::ContentFilterUnblockHandler::requestUnblockAsync const):
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
(WebCore::NetworkExtensionContentFilter::create):
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
(WebCore::ParentalControlsContentFilter::enabled const):
(WebCore::ParentalControlsContentFilter::create):
(WebCore::ParentalControlsContentFilter::ParentalControlsContentFilter):
(WebCore::ParentalControlsContentFilter::responseReceived):
(WebCore::ParentalControlsContentFilter::addData):
(WebCore::ParentalControlsContentFilter::finishedAddingData):
(WebCore::ParentalControlsContentFilter::unblockHandler const):
(WebCore::ParentalControlsContentFilter::updateFilterState):
(WebCore::ParentalControlsContentFilter::enabled): Deleted.
(WebCore::ParentalControlsContentFilter::setUsesWebContentRestrictions): Deleted.
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h: Copied from Source/WebCore/platform/cocoa/PhotosFormatSoftLink.mm.
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm: Added.
(WebCore::wcrBrowserEngineClientEnabled):
(WebCore::ParentalControlsURLFilter::singleton):
(WebCore::ParentalControlsURLFilter::ParentalControlsURLFilter):
(WebCore::ParentalControlsURLFilter::isEnabled const):
(WebCore::ParentalControlsURLFilter::isURLAllowed):
(WebCore::ParentalControlsURLFilter::allowURL):
* Source/WebCore/platform/cocoa/PhotosFormatSoftLink.mm:
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm:
(WebCore::CDMPrivateMediaSourceAVFObjC::cdm const):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.mm:
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.mm:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm:
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.mm:
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h:
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.h:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
* Source/WebCore/platform/graphics/mac/controls/SearchControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/SearchControlMac.mm:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
* Source/WebCore/testing/MockContentFilter.cpp:
(WebCore::MockContentFilter::create):
* Source/WebCore/testing/MockContentFilter.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startContentFiltering):
(WebKit::NetworkResourceLoader::usesWebContentRestrictions):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:

Canonical link: https://commits.webkit.org/292963@main
